src/scanner.c

```diff
@@ -0,0 +1,163 @@
+#include "tree_sitter/parser.h"
+#include "tree_sitter/alloc.h"
+#include <stdbool.h>
+#include <stdint.h>
+
+#define skip lexer->advance(lexer, true)
+#define adv lexer->advance(lexer, false)
+#define eof lexer->eof(lexer)
+
+enum TokenType {
+    RAW_STRING_BEGIN,
+    RAW_STRING_CONTENT,
+    RAW_STRING_END,
+    ERROR_SENTINEL,
+};
+
+typedef struct {
+    uint8_t level;
+} Scanner;
+
+static uint8_t consume_chars(TSLexer *lexer, char c) {
+    uint8_t count = 0;
+    while (lexer->lookahead == c && !eof) {
+        adv;
+        ++count;
+    }
+    return count;
+}
+
+static uint32_t consume_until(TSLexer *lexer, char c) {
+    uint32_t count = 0;
+    while (lexer->lookahead != c && !eof) {
+        adv;
+        count++;
+    }
+    return count;
+}
+
+
+void *tree_sitter_nu_external_scanner_create(void) {
+    Scanner *scanner = ts_malloc(sizeof(Scanner));
+    scanner->level = 0;
+    return scanner;
+}
+
+
+void tree_sitter_nu_external_scanner_destroy(void *payload) {
+    ts_free(payload);
+}
+
+
+unsigned tree_sitter_nu_external_scanner_serialize(
+    void *payload,
+    char *buffer
+) {
+    Scanner *s = (Scanner *) payload;
+    buffer[0] = s->level;
+    return 1;
+}
+
+
+void tree_sitter_nu_external_scanner_deserialize(
+    void *payload,
+    const char *buffer,
+    unsigned length
+) {
+    Scanner *s = (Scanner *) payload;
+    s->level = 0;
+    if (length == 1) {
+        s->level = buffer[0];
+    }
+}
+
+static void skip_whitespace(TSLexer *lexer) {
+    while ((lexer->lookahead ==  ' ' || lexer->lookahead == '\t'
+            || lexer->lookahead == '\n') && !eof) {
+        skip;
+    }
+}
+
+static bool scan_raw_string_begin(TSLexer *lexer, Scanner *s) {
+    lexer->log(lexer, "BEGIN\n");
+    // scan for r#' r##' or more #
+    skip_whitespace(lexer);
+
+    if (lexer->lookahead == 'r') {
```

Comment:
This check is already done.

Reply:
This also makes `skip_whitespace` obsolete, thanks!